fccodegen/tests/test_printer.py

```diff
@@ -0,0 +1,197 @@
+    assert p.doprint(sp.atanh(x)) == 'math.atanh(x)'
+    assert p.doprint(sp.ceiling(x)) == 'math.ceil(x)'
+    assert p.doprint(sp.cos(x)) == 'math.cos(x)'
+    assert p.doprint(sp.cosh(x)) == 'math.cosh(x)'
+    assert p.doprint(sp.exp(x)) == 'math.exp(x)'
+    assert p.doprint(sp.factorial(x)) == 'math.factorial(x)'
+    assert p.doprint(sp.floor(x)) == 'math.floor(x)'
+    assert p.doprint(sp.log(x)) == 'math.log(x)'
+    assert p.doprint(sp.sin(x)) == 'math.sin(x)'
+    assert p.doprint(sp.sinh(x)) == 'math.sinh(x)'
+    assert p.doprint(sp.tan(x)) == 'math.tan(x)'
+    assert p.doprint(sp.tanh(x)) == 'math.tanh(x)'
+
+    # Conditions
+    assert p.doprint(sp.Eq(x, y)) == 'x == y'
+    assert p.doprint(sp.Eq(x, sp.Eq(y, z))) == 'x == (y == z)'
+    assert p.doprint(sp.Eq(sp.Eq(x, y), z)) == '(x == y) == z'
+    assert p.doprint(sp.Ne(x, y)) == 'x != y'
+    assert p.doprint(sp.Gt(x, y)) == 'x > y'
+    assert p.doprint(sp.Lt(x, y)) == 'x < y'
+    assert p.doprint(sp.Ge(x, y)) == 'x >= y'
+    assert p.doprint(sp.Le(x, y)) == 'x <= y'
+    assert p.doprint(sp.Eq(sp.Eq(x, 3), 12)) == '(x == 3) == 12'
+
+    # Boolean logic
+    assert p.doprint(True) == 'True'
+    assert p.doprint(False) == 'False'
+    assert p.doprint(sp.Eq(x, x)) == 'True'
+    assert p.doprint(sp.Ne(x, x)) == 'False'
+    assert p.doprint(sp.And(sp.Eq(x, y), sp.Eq(x, z))) == 'x == y and x == z'
+    assert (
+        p.doprint(sp.And(sp.Eq(x, y), sp.Eq(x, z), sp.Eq(x, 2))) ==
+        'x == 2 and x == y and x == z')
+    assert p.doprint(sp.Or(sp.Eq(x, y), sp.Eq(x, z))) == 'x == y or x == z'
+    assert (
+        p.doprint(sp.Or(sp.Eq(x, y), sp.Eq(x, z), sp.Eq(x, 2))) ==
+        'x == 2 or x == y or x == z')
+    a, b, c = x > 2, x > y, x > z
+    assert p.doprint(a & b) == 'x > 2 and x > y'
+    # 1 or (0 and 0) = 1 = 1 or 0 and 0 -- and binds stronger
+    # (1 or 0) and 0 = 0
+    assert p.doprint(a | (b & c)) == 'x > 2 or x > y and x > z'
+    assert p.doprint((a | b) & c) == 'x > z and (x > 2 or x > y)'
+
+    # Piecewise expressions
+    e = sp.Piecewise((0, x > 0), (1, True))
+    assert p.doprint(e) == '((0) if (x > 0) else (1))'
+    e = sp.Piecewise((0, x > 0), (1, x > 1), (2, True))
+    assert p.doprint(e) == '((0) if (x > 0) else ((1) if (x > 1) else (2)))'
+    e = sp.Piecewise((0, x > 0), (1, x > 1), (2, True), (3, x > 3))
+    assert p.doprint(e) == '((0) if (x > 0) else ((1) if (x > 1) else (2)))'
+    # Sympy filters out False statements
+    e = sp.Piecewise(
+        (0, x > 0), (1, x != x), (2, True), (3, x > 3),
+        evaluate=False)
+    assert p.doprint(e) == '((0) if (x > 0) else (2))'
+
+    # Longer expressions
+    assert (
+        p.doprint((x + y) / (2 + z / sp.exp(x - y))) ==
+        '(x + y) / (2 + z * math.exp(y - x))')
+    assert p.doprint((y + sp.sin(x))**-1) == '1 / (y + math.sin(x))'
+
+    # Unsupported sympy item
+    e = sp.Matrix()
+    with pytest.raises(RuntimeError):
```

Comment:
Can we have a more specific exception thrown please? `ValueError` might be a good candidate.

Reply:
Done!